app/src/db/bulk_ops.py

```diff
@@ -0,0 +1,150 @@
+"""Bulk database operations for performance.
+
+Provides a bulk_upsert function
+"""
+from typing import Any, Optional, Sequence
+
+import psycopg
+from psycopg import rows, sql
+
+Connection = psycopg.Connection
+Cursor = psycopg.Cursor
+kwargs_row = rows.kwargs_row
+
+
+def bulk_upsert(
+    cur: psycopg.Cursor,
+    table: str,
+    attributes: Sequence[str],
+    objects: Sequence[Any],
+    constraint: str,
+    update_condition: Optional[sql.SQL] = None,
+) -> None:
+    """Bulk insert or update a sequence of objects.
+
+    Insert a sequence of objects, or update on conflict.
+    Write data from one table to another.
+    If there are conflicts due to unique constraints, overwrite existing data.
+
+    Args:
+      cur: the Cursor object from the pyscopg library
+      table: the name of the table to insert into or update
+      attributes: a sequence of attribute names to copy from each object
+      objects: a sequence of objects to upsert
+      constraint: the table unique constraint to use to determine conflicts
+      update_condition: optional WHERE clause to limit updates for a
+        conflicting row
+    """
+    if not update_condition:
+        update_condition = sql.SQL("")
+
+    temp_table = f"temp_{table}"
+    create_temp_table(cur, temp_table=temp_table, src_table=table)
```

Comment:
This is probably a very niche edge case, but what would happen if two temp tables were created with the same name by different processes? Does that cause any issues, or does them being in the transactions entirely shield them?

Reply:
Great question, I tested it locally and it looks like the transaction isolation works like you'd expect. Here's the SQL I ran:

```BEGIN;
CREATE TEMP TABLE test (id INT) ON COMMIT DROP;
SELECT * FROM test;

-- In a separate connection!
BEGIN;
CREATE TEMP TABLE test (other INT) ON COMMIT DROP;
SELECT * FROM test;
COMMIT;

-- Back in the original connection
COMMIT;```